### PR TITLE
🦺 718 - add granular errors for password requirements

### PR DESCRIPTION
**Pull request by anncatton**
## Description of Changes

Adds separate validations for all password requirements + sets up Register Step 2 to display multiple errors on `password` field.
PR # 1 of Password changes for #718 + #726 

### Consent UI
- adds `criteriaMode: "all"` to support multiple errors on a single field in the react hook form. This adds a `types` property to formState `errors`, where multiple errors can be accessed. Inputs that are intended to display just one error at a time can still rely on `errors[field].type` and `errors[field].message`

### Types
- adds regexes to support separate validations for each Password requirement: one lowercase, one uppercase, correct character length, and no whitespaces. Symbols and numbers will be allowed characters.
- adds validation function to check for each Password requirement separately, for use in the UI form. Will be used with `superrefine()` to allow multiple validation errors to be added to Zod context
- adds tests for separate + multiple password validations
- updates existing tests for Registration to match Password regex pattern
- **Note**: `confirmPassword` will be removed in https://github.com/OHCRN/platform/issues/726

### Dependencies
- minor version bumps for `zod` and `@hookform/resolvers`

### Special Instructions
Run `pnpm i` to pick up dependency updates.
Run `pnpm run build` from `types` package to pick up Password type changes.

## PR Readiness Checklist

- [x] "Expected Outcome(s)" in ticket have been met
- [x] Ticket number included in PR title
- [x] Connected ticket to PR
- [x] Labels added to PR for service name (`consent-api`, `data-mapper`, etc...), type (`chore`, `documentation`, etc...), status (`draft`, `on-hold`, etc...) **if applicable**
- [x] Manual testing completed
- [x] Builds locally without errors or warnings
- [x] Tests are updated (if required) and passing
- [ ] PR feedback has been addressed **if applicable**
- [ ] New environment variables added to `.env.schema` files, `README.md`
- [ ] Added copyrights to new files

### CODE REVIEW

File: apps/consent-ui/src/components/common/Form/fieldsets/FieldSet/index.tsx [hunk 1]
```diff
@@ -68,7 +68,12 @@ const FieldSet = ({
 
 			<div className={styles.inputGridArea}>
 				{children}
-				{error && <InputError>{error}</InputError>}
+
+				{Array.isArray(error) ? (
```
Comment (anncatton) [inline apps/consent-ui/src/components/common/Form/fieldsets/FieldSet/index.tsx:72]: This change is reverted in https://github.com/OHCRN/platform/pull/732

File: packages/types/tests/services/participantRegistrationRefinements.spec.ts [hunk 2]
```diff
@@ -458,4 +460,115 @@ describe('ParticipantRegistrationRequest', () => {
...
+			it('Returns true if all password requirements are met', () => {
+				const result = MockParticipantRegistrationRequest.safeParse(adultConsentTestData);
+				expect(result.success).true;
+			});
+
+			it('Adds invalid character error when it contains whitespace', () => {
+				const result = MockParticipantRegistrationRequest.safeParse({
+					...adultConsentTestData,
+					password: 'My password has SPaces',
+					confirmPassword: 'My password has SPaces'
```
Comment (anncatton) [inline packages/types/tests/services/participantRegistrationRefinements.spec.ts:475]: `confirmPassword` added in these tests so they pass, but is removed in #726

File: packages/types/src/services/consentUi/requests/Register.ts [hunk 3]
```diff
@@ -77,12 +83,43 @@ const RegisterEmailAddressFieldsRefined = RegisterRequestEmailAddressFields.supe
...
+ * Allows for multiple errors to be added to Zod refinement context for password field
+ * Test for:
+ * - contains no disallowed characters (whitespace)
+ * - has at least one lowercase letter
+ * - has at least one uppercase letter
+ * - has required number of characters
+ * @param props { password: string }
+ * @param ctx z.RefinementCtx
+ * @returns void
+ */
```
Comment (mistryrn) [inline packages/types/src/services/consentUi/requests/Register.ts:97]: This is great!

**Review by mistryrn — COMMENTED** [review]
Looks great! Will re-review once the whitespace regex and tests are updated 🙂

File: packages/types/src/entities/fields/Password.ts [hunk 4]
```diff
@@ -19,7 +19,32 @@
 
 import { z } from 'zod';
 
-import { PASSWORD_REGEX } from '../../common/regexes.js';
+import { TrimmedString } from '../../common/String.js';
+import {
+	PASSWORD_MAXIMUM_LENGTH,
+	PASSWORD_MININUM_LENGTH,
```
Comment (samrichca) [inline packages/types/src/entities/fields/Password.ts:25]: typo: 
```suggestion
	PASSWORD_MINIMUM_LENGTH,
```

File: packages/types/src/services/consentUi/requests/Register.ts [hunk 5]
```diff
@@ -77,12 +83,43 @@ const RegisterEmailAddressFieldsRefined = RegisterRequestEmailAddressFields.supe
...
+	if (!LengthRequirement.safeParse(password).success) {
+		addZodCustomError(ctx, path, 'invalidPasswordLength');
+	}
+};
+
 const PasswordFields = z
 	.object({
-		confirmPassword: Password,
-		password: Password,
+		confirmPassword: TrimmedString,
```
Comment (samrichca) [inline packages/types/src/services/consentUi/requests/Register.ts:117]: trimming is getting removed right?

Comment (anncatton) [reply packages/types/src/services/consentUi/requests/Register.ts:117]: yes ma'am, in the 3rd pr

File: packages/types/src/entities/fields/Password.ts (hunk 4 above)
Comment (anncatton) [reply packages/types/src/entities/fields/Password.ts:25]: lol good catch, i'll fix this in the last pr 🙏

**Review by mistryrn — APPROVED** [review]
🚀 🔐

**Review by mistryrn — APPROVED** [review]

**Pull request merged**